src/account_id_ref.rs

```diff
@@ -0,0 +1,691 @@
+/// and works quite similarly to [`Path`]. Like with [`str`] and [`Path`], you
+/// can't have a value of type `AccountIdRef`, but you can have a reference like `&AccountIdRef` or
+/// `&mut AccountIdRef`.
+///
+/// This type supports zero-copy deserialization offered by [`serde`](https://docs.rs/serde/), but cannot
+/// do the same for [`borsh`](https://docs.rs/borsh/) since the latter does not support zero-copy.
+///
+/// # Examples
+/// ```
+/// use near_account_id::{AccountId, AccountIdRef};
+/// use std::convert::{TryFrom, TryInto};
+///
+/// // Construction
+/// let alice = AccountIdRef::new("alice.near").unwrap();
+/// assert!(AccountIdRef::new("invalid.").is_err());
+///
+/// // Initialize without validating
+/// let alice_unchecked = AccountIdRef::new_unchecked("alice.near");
+/// assert_eq!(alice, alice_unchecked);
+/// ```
+///
+/// [`FromStr`]: std::str::FromStr
+/// [`Path`]: std::path::Path
+#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Hash)]
+#[cfg_attr(feature = "abi", derive(schemars::JsonSchema, BorshSchema))]
+pub struct AccountIdRef(pub(crate) str);
+
+impl AccountIdRef {
+    /// Shortest valid length for a NEAR Account ID.
+    pub const MIN_LEN: usize = crate::validation::MIN_LEN;
+    /// Longest valid length for a NEAR Account ID.
+    pub const MAX_LEN: usize = crate::validation::MAX_LEN;
+
+    /// Construct a [`&AccountIdRef`](AccountIdRef) from a string reference.
+    ///
+    /// This constructor validates the provided ID, and will produce an error when validation fails.
+    pub fn new<S: AsRef<str> + ?Sized>(id: &S) -> Result<&Self, ParseAccountError> {
+        let id = id.as_ref();
+        validate(id)?;
+
+        // Safety:
+        // - a newtype struct is guaranteed to have the same memory layout as its only field
+        // - the borrow checker will enforce its rules appropriately on the resulting reference
+        Ok(unsafe { &*(id as *const str as *const Self) })
+    }
+
+    /// Construct a [`&AccountIdRef`](AccountIdRef) from a string reference without validating the address.
+    /// It is the responsibility of the caller to ensure the account ID is valid.
+    ///
+    /// For more information, read: <https://docs.near.org/docs/concepts/account#account-id-rules>
+    pub fn new_unchecked<S: AsRef<str> + ?Sized>(id: &S) -> &Self {
+        let id = id.as_ref();
+        debug_assert!(validate(id).is_ok());
+
+        // Safety: see `AccountId::new`
+        unsafe { &*(id as *const str as *const Self) }
+    }
+
+    /// Construct a [`&mut AccountIdRef`](AccountIdRef) from a mutable string reference.
+    ///
+    /// This constructor validates the provided ID and will produce an error when validation fails.
+    pub fn new_mut<S: AsMut<str> + ?Sized>(id: &mut S) -> Result<&mut Self, ParseAccountError> {
+        let id = id.as_mut();
+        validate(id)?;
+
+        // Safety: see `AccountId::new`
+        Ok(unsafe { &mut *(id as *mut str as *mut Self) })
+    }
```

Comment:
Ah, that's fair. I haven't thought that through - I was mostly copying what looked useful from the `Path`/`PathBuf` API.

I suppose this could be used for things like [std::mem::swap](https://doc.rust-lang.org/std/mem/fn.swap.html), but that's unlikely to be useful. I'm happy to remove it.